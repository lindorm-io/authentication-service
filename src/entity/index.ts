export * from "./Account";
export * from "./RequestLimit";
export * from "./Session";
