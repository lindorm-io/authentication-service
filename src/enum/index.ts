export * from "./account-event";
export * from "./audience";
export * from "./device-event";
export * from "./grant-type";
export * from "./mongo-collection";
export * from "./multi-factor-challenge-type";
export * from "./response-type";
export * from "./session-event";
