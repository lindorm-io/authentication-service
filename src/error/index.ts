export * from "./AccountNotFoundError";
export * from "./AssertClientSecretError";
export * from "./AssertCodeChallengeError";
export * from "./AssertDeviceChallengeError";
export * from "./AssertOTPError";
export * from "./DeviceNotFoundError";
export * from "./ExpiredSessionError";
export * from "./InvalidAuthorizationError";
export * from "./InvalidAuthorizationTokenError";
export * from "./InvalidClientError";
export * from "./InvalidDeviceError";
export * from "./InvalidExpiryString";
export * from "./InvalidGrantTypeError";
export * from "./InvalidPermissionError";
export * from "./InvalidRefreshTokenError";
export * from "./InvalidScopeError";
export * from "./InvalidSubjectError";
export * from "./LockedAccountError";
export * from "./RejectedClientError";
export * from "./RequestLimitBackOffError";
