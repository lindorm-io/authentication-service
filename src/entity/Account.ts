import { AccountEvent } from "../enum";
import { EntityBase, EntityCreationError, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { Permission } from "@lindorm-io/jwt";

export interface IAccountOTP {
  signature: string;
  uri: string;
}

export interface IAccountPassword {
  signature: string;
  updated: Date;
}

export interface IAccount extends IEntity {
  email: string;
  identityLinked: boolean;
  otp: IAccountOTP;
  password: IAccountPassword;
  permission: string;
}

export interface IAccountOptions extends IEntityBaseOptions {
  email: string;
  identityLinked?: boolean;
  otp?: IAccountOTP;
  password?: IAccountPassword;
  permission?: string;
}

export class Account extends EntityBase implements IAccount {
  private _email: string;
  private _identityLinked: boolean;
  private _otp: IAccountOTP;
  private _password: IAccountPassword;
  private _permission: string;

  constructor(options: IAccountOptions) {
    super(options);
    this._email = options.email.toLowerCase();
    this._identityLinked = options.identityLinked || false;
    this._otp = {
      signature: options.otp?.signature || null,
      uri: options.otp?.uri || null,
    };
    this._password = {
      signature: options.password?.signature || null,
      updated: options.password?.updated || null,
    };
    this._permission = options.permission || Permission.USER;
  }

  public get email(): string {
    return this._email;
  }
  public set email(email: string) {
    this._email = email.toLowerCase();
    this.addEvent(AccountEvent.EMAIL_CHANGED, { email: this._email });
  }

  public get identityLinked(): boolean {
    return this._identityLinked;
  }
  public set identityLinked(identityLinked: boolean) {
    this._identityLinked = identityLinked;
    this.addEvent(AccountEvent.EMAIL_CHANGED, { identityLinked: this._identityLinked });
  }

  public get otp(): IAccountOTP {
    return this._otp;
  }
  public set otp(otp: IAccountOTP) {
    this._otp = otp;
    this.addEvent(AccountEvent.OTP_CHANGED, { otp: this._otp });
  }

  public get password(): IAccountPassword {
    return this._password;
  }
  public set password(password: IAccountPassword) {
    this._password = password;
    this.addEvent(AccountEvent.PASSWORD_CHANGED, { password: this._password });
  }

  public get permission(): string {
    return this._permission;
  }
  public set permission(permission: string) {
    this._permission = permission;
    this.addEvent(AccountEvent.PERMISSION_CHANGED, { permission: this._permission });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== AccountEvent.CREATED) continue;
      throw new EntityCreationError("Account");
    }
    this.addEvent(AccountEvent.CREATED, {
      email: this._email,
      permission: this._permission,
      password: this._password,
      otp: this._otp,
      created: this._created,
      updated: this._updated,
    });
  }
}
