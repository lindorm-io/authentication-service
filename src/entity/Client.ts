import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { ClientEvent } from "../enum";

export interface IClient extends IEntity {
  approved: boolean;
  description: string;
  emailAuthorizationUri: string;
  name: string;
  secret: string;
}

export interface IClientOptions extends IEntityBaseOptions {
  approved?: boolean;
  description?: string;
  emailAuthorizationUri?: string;
  name?: string;
  secret?: string;
}

export class Client extends EntityBase implements IClient {
  private _approved: boolean;
  private _description: string;
  private _emailAuthorizationUri: string;
  private _name: string;
  private _secret: string;

  constructor(options?: IClientOptions) {
    super(options);
    this._approved = options?.approved || false;
    this._description = options?.description || null;
    this._emailAuthorizationUri = options?.emailAuthorizationUri || null;
    this._name = options?.name || null;
    this._secret = options?.secret || null;
  }

  public get approved(): boolean {
    return this._approved;
  }
  public set approved(approved: boolean) {
    this._approved = approved;
    this.addEvent(ClientEvent.APPROVED_CHANGED, { approved: this._approved });
  }

  public get emailAuthorizationUri(): string {
    return this._emailAuthorizationUri;
  }
  public set emailAuthorizationUri(uri: string) {
    this._emailAuthorizationUri = uri;
    this.addEvent(ClientEvent.EMAIL_AUTHORIZATION_URI_CHANGED, { uri: this._emailAuthorizationUri });
  }

  public get description(): string {
    return this._description;
  }
  public set description(description: string) {
    this._description = description;
    this.addEvent(ClientEvent.DESCRIPTION_CHANGED, { description: this._description });
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(ClientEvent.NAME_CHANGED, { name: this._name });
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(ClientEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== ClientEvent.CREATED) continue;
      throw new Error("Client has already been created");
    }
    this.addEvent(ClientEvent.CREATED, {
      name: this._name,
      emailAuthorizationUri: this._emailAuthorizationUri,
      approved: this._approved,
      description: this._description,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
