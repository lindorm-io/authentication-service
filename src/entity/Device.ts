import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/core";
import { DeviceEvent } from "../enum";

export interface IDevicePIN {
  signature: string;
  updated: Date;
}

export interface IDevice extends IEntity {
  accountId: string;
  name: string;
  pin: IDevicePIN;
  publicKey: string;
  secret: string;
}

export interface IDeviceOptions extends IEntityBaseOptions {
  accountId: string;
  name?: string;
  pin?: IDevicePIN;
  publicKey: string;
  secret?: string;
}

export class Device extends EntityBase implements IDevice {
  private _accountId: string;
  private _name: string;
  private _pin: IDevicePIN;
  private _publicKey: string;
  private _secret: string;

  constructor(options: IDeviceOptions) {
    super(options);
    this._accountId = options.accountId;
    this._name = options.name || null;
    this._pin = {
      signature: options.pin?.signature || null,
      updated: options.pin?.updated || null,
    };
    this._publicKey = options.publicKey;
    this._secret = options.secret || null;
  }

  public get accountId(): string {
    return this._accountId;
  }

  public get name(): string {
    return this._name;
  }
  public set name(name: string) {
    this._name = name;
    this.addEvent(DeviceEvent.NAME_CHANGED, { name: this._name });
  }

  public get pin(): IDevicePIN {
    return this._pin;
  }
  public set pin(pin: IDevicePIN) {
    this._pin = pin;
    this.addEvent(DeviceEvent.PIN_CHANGED, { pin: this._pin });
  }

  public get publicKey(): string {
    return this._publicKey;
  }

  public get secret(): string {
    return this._secret;
  }
  public set secret(secret: string) {
    this._secret = secret;
    this.addEvent(DeviceEvent.SECRET_CHANGED, { secret: this._secret });
  }

  public create(): void {
    for (const evt of this._events) {
      if (evt.name !== DeviceEvent.CREATED) continue;
      throw new Error("Device has already been created");
    }
    this.addEvent(DeviceEvent.CREATED, {
      accountId: this._accountId,
      name: this._name,
      pin: this._pin,
      publicKey: this._publicKey,
      secret: this._secret,
      created: this._created,
      updated: this._updated,
    });
  }
}
