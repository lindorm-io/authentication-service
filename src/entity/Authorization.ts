import { EntityBase, IEntity, IEntityBaseOptions } from "@lindorm-io/entity";
import { GrantType } from "../enum";
import { Scope } from "@lindorm-io/jwt";

export interface IAuthorization extends IEntity {
  challengeId: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export interface IAuthorizationOptions extends IEntityBaseOptions {
  challengeId?: string;
  clientId: string;
  codeChallenge: string;
  codeMethod: string;
  deviceId?: string;
  email: string;
  expires: Date;
  grantType: GrantType;
  otpCode?: string;
  redirectUri: string;
  responseType: string;
  scope: Array<Scope>;
}

export class Authorization extends EntityBase implements IAuthorization {
  readonly challengeId: string;
  readonly clientId: string;
  readonly codeChallenge: string;
  readonly codeMethod: string;
  readonly deviceId: string;
  readonly email: string;
  readonly expires: Date;
  readonly grantType: GrantType;
  readonly otpCode: string;
  readonly redirectUri: string;
  readonly responseType: string;
  readonly scope: Array<Scope>;

  constructor(options: IAuthorizationOptions) {
    super(options);

    this.challengeId = options.challengeId || null;
    this.clientId = options.clientId;
    this.codeChallenge = options.codeChallenge;
    this.codeMethod = options.codeMethod;
    this.deviceId = options.deviceId || null;
    this.email = options.email;
    this.expires = options.expires;
    this.grantType = options.grantType;
    this.otpCode = options.otpCode || null;
    this.redirectUri = options.redirectUri;
    this.responseType = options.responseType;
    this.scope = options.scope;
  }

  public create(): void {
    /* intentionally left empty */
  }
}
