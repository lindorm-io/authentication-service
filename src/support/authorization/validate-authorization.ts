import { IKoaAuthContext } from "../../typing";
import { Authorization } from "../../entity";
import { assertAuthorizationIsNotExpired } from "./expires";
import { InvalidClientError } from "@lindorm-io/koa-client";
import { InvalidDeviceError, InvalidGrantTypeError, InvalidSubjectError } from "../../error";
import { assertCodeChallenge } from "./code-challenge";

export interface IValidateAuthorizationOptions {
  codeVerifier: string;
  email: string;
  grantType: string;
}

export const validateAuthorization = (ctx: IKoaAuthContext) => async (
  options: IValidateAuthorizationOptions,
): Promise<Authorization> => {
  const { cache, client, metadata, token } = ctx;
  const { codeVerifier, email, grantType } = options;
  const {
    authorization: { id: authorizationId },
  } = token;

  const authorization = await cache.authorization.find(authorizationId);

  try {
    assertAuthorizationIsNotExpired(authorization);

    if (authorization.clientId !== client.id) {
      throw new InvalidClientError(client.id);
    }

    if (authorization.deviceId !== metadata.deviceId) {
      throw new InvalidDeviceError(metadata.deviceId);
    }

    if (authorization.email !== email) {
      throw new InvalidSubjectError(email);
    }

    if (authorization.grantType !== grantType) {
      throw new InvalidGrantTypeError(grantType);
    }

    assertCodeChallenge(authorization, codeVerifier);

    return authorization;
  } catch (err) {
    await cache.authorization.remove(authorization);

    throw err;
  }
};
