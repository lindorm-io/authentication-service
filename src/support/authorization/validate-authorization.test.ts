import MockDate from "mockdate";
import { Authorization } from "../../entity";
import { CacheEntityNotFoundError } from "@lindorm-io/redis";
import { Client, InvalidClientError } from "@lindorm-io/koa-client";
import { GrantType } from "../../enum";
import { InvalidDeviceError, InvalidGrantTypeError, InvalidSubjectError } from "../../error";
import { getTestAuthorization, getTestCache, getTestClient, inMemoryCache, resetStore } from "../../test";
import { validateAuthorization } from "./validate-authorization";

jest.mock("uuid", () => ({
  v4: jest.fn(() => "be3a62d1-24a0-401c-96dd-3aff95356811"),
}));
jest.mock("./code-challenge", () => ({
  assertCodeChallenge: jest.fn(),
}));
jest.mock("./expires", () => ({
  assertAuthorizationIsNotExpired: jest.fn(),
}));

MockDate.set("2020-01-01 08:00:00.000");

describe("validateAuthorization", () => {
  let ctx: any;
  let client: Client;
  let authorization: Authorization;

  beforeEach(async () => {
    client = getTestClient();
    authorization = getTestAuthorization({});

    ctx = {
      client: client,
      metadata: { deviceId: authorization.deviceId },
      cache: await getTestCache(),
      token: { authorization: { id: authorization.id } },
    };

    await ctx.cache.authorization.create(authorization);
  });

  afterEach(resetStore);

  test("should find and validate authorization", async () => {
    await expect(
      validateAuthorization(ctx)({
        codeVerifier: "MHNnb24yMzBqZ2hvNDJuMDI5ajAyNGdt",
        grantType: GrantType.EMAIL_OTP,
        email: "email@lindorm.io",
      }),
    ).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should throw error when authorizationId is wrong", async () => {
    ctx.token.authorization.id = "wrong";

    await expect(
      validateAuthorization(ctx)({
        codeVerifier: "MHNnb24yMzBqZ2hvNDJuMDI5ajAyNGdt",
        grantType: GrantType.EMAIL_OTP,
        email: "email@lindorm.io",
      }),
    ).rejects.toStrictEqual(expect.any(CacheEntityNotFoundError));

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should throw error when clientId is wrong", async () => {
    ctx.client = new Client({ id: "wrong" });

    await expect(
      validateAuthorization(ctx)({
        codeVerifier: "MHNnb24yMzBqZ2hvNDJuMDI5ajAyNGdt",
        grantType: GrantType.EMAIL_OTP,
        email: "email@lindorm.io",
      }),
    ).rejects.toStrictEqual(expect.any(InvalidClientError));

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should throw error when deviceId is wrong", async () => {
    ctx.metadata.deviceId = "wrong";

    await expect(
      validateAuthorization(ctx)({
        codeVerifier: "MHNnb24yMzBqZ2hvNDJuMDI5ajAyNGdt",
        grantType: GrantType.EMAIL_OTP,
        email: "email@lindorm.io",
      }),
    ).rejects.toStrictEqual(expect.any(InvalidDeviceError));

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should throw error when grantType is wrong", async () => {
    await expect(
      validateAuthorization(ctx)({
        codeVerifier: "MHNnb24yMzBqZ2hvNDJuMDI5ajAyNGdt",
        grantType: GrantType.PASSWORD,
        email: "email@lindorm.io",
      }),
    ).rejects.toStrictEqual(expect.any(InvalidGrantTypeError));

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should throw error when email is wrong", async () => {
    await expect(
      validateAuthorization(ctx)({
        codeVerifier: "MHNnb24yMzBqZ2hvNDJuMDI5ajAyNGdt",
        grantType: GrantType.EMAIL_OTP,
        email: "wrong",
      }),
    ).rejects.toStrictEqual(expect.any(InvalidSubjectError));

    expect(inMemoryCache).toMatchSnapshot();
  });
});
