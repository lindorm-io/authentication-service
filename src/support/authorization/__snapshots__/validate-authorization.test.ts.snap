// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validateAuthorization should find and validate authorization 1`] = `
Authorization {
  "_created": 2020-01-01T07:00:00.000Z,
  "_events": Array [],
  "_id": "4923fabc-aab2-4804-b92b-2aa96c4999a1",
  "_updated": 2020-01-01T07:00:00.000Z,
  "_version": 0,
  "challengeId": "1997d969-0e46-41c5-8c5e-836d643e78c9",
  "clientId": "9ba5e0c5-f2aa-4706-a27d-66369b547b0b",
  "codeChallenge": "H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=",
  "codeMethod": "sha256",
  "deviceId": null,
  "email": "email@lindorm.io",
  "expires": 2099-01-01T00:00:00.000Z,
  "grantType": "email_otp",
  "otpCode": null,
  "redirectUri": "https://redirect.uri/",
  "responseType": "code",
  "scope": Array [
    "default",
    "edit",
    "openid",
  ],
}
`;

exports[`validateAuthorization should find and validate authorization 2`] = `
Object {
  "Authorization::4923fabc-aab2-4804-b92b-2aa96c4999a1": Object {
    "blob": "{\\"json\\":{\\"id\\":\\"4923fabc-aab2-4804-b92b-2aa96c4999a1\\",\\"version\\":\\"0\\",\\"created\\":\\"2020-01-01T07:00:00.000Z\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\",\\"events\\":[],\\"challengeId\\":\\"1997d969-0e46-41c5-8c5e-836d643e78c9\\",\\"clientId\\":\\"9ba5e0c5-f2aa-4706-a27d-66369b547b0b\\",\\"codeChallenge\\":\\"H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=\\",\\"codeMethod\\":\\"sha256\\",\\"deviceId\\":\\"null\\",\\"email\\":\\"email@lindorm.io\\",\\"expires\\":\\"2099-01-01T00:00:00.000Z\\",\\"grantType\\":\\"email_otp\\",\\"otpCode\\":\\"null\\",\\"redirectUri\\":\\"https://redirect.uri/\\",\\"responseType\\":\\"code\\",\\"scope\\":[\\"default\\",\\"edit\\",\\"openid\\"]},\\"meta\\":{\\"id\\":\\"string\\",\\"version\\":\\"number\\",\\"created\\":\\"date\\",\\"updated\\":\\"date\\",\\"events\\":[],\\"challengeId\\":\\"string\\",\\"clientId\\":\\"string\\",\\"codeChallenge\\":\\"string\\",\\"codeMethod\\":\\"string\\",\\"deviceId\\":\\"null\\",\\"email\\":\\"string\\",\\"expires\\":\\"date\\",\\"grantType\\":\\"string\\",\\"otpCode\\":\\"null\\",\\"redirectUri\\":\\"string\\",\\"responseType\\":\\"string\\",\\"scope\\":[\\"string\\",\\"string\\",\\"string\\"]}}",
    "expiresInSeconds": null,
  },
}
`;

exports[`validateAuthorization should throw error when authorizationId is wrong 1`] = `
Object {
  "Authorization::4923fabc-aab2-4804-b92b-2aa96c4999a1": Object {
    "blob": "{\\"json\\":{\\"id\\":\\"4923fabc-aab2-4804-b92b-2aa96c4999a1\\",\\"version\\":\\"0\\",\\"created\\":\\"2020-01-01T07:00:00.000Z\\",\\"updated\\":\\"2020-01-01T07:00:00.000Z\\",\\"events\\":[],\\"challengeId\\":\\"1997d969-0e46-41c5-8c5e-836d643e78c9\\",\\"clientId\\":\\"9ba5e0c5-f2aa-4706-a27d-66369b547b0b\\",\\"codeChallenge\\":\\"H4LnTn7e1DltMsohJgIeKSNgpvppJ1qP6QRRD9Ai1pw=\\",\\"codeMethod\\":\\"sha256\\",\\"deviceId\\":\\"null\\",\\"email\\":\\"email@lindorm.io\\",\\"expires\\":\\"2099-01-01T00:00:00.000Z\\",\\"grantType\\":\\"email_otp\\",\\"otpCode\\":\\"null\\",\\"redirectUri\\":\\"https://redirect.uri/\\",\\"responseType\\":\\"code\\",\\"scope\\":[\\"default\\",\\"edit\\",\\"openid\\"]},\\"meta\\":{\\"id\\":\\"string\\",\\"version\\":\\"number\\",\\"created\\":\\"date\\",\\"updated\\":\\"date\\",\\"events\\":[],\\"challengeId\\":\\"string\\",\\"clientId\\":\\"string\\",\\"codeChallenge\\":\\"string\\",\\"codeMethod\\":\\"string\\",\\"deviceId\\":\\"null\\",\\"email\\":\\"string\\",\\"expires\\":\\"date\\",\\"grantType\\":\\"string\\",\\"otpCode\\":\\"null\\",\\"redirectUri\\":\\"string\\",\\"responseType\\":\\"string\\",\\"scope\\":[\\"string\\",\\"string\\",\\"string\\"]}}",
    "expiresInSeconds": null,
  },
}
`;

exports[`validateAuthorization should throw error when clientId is wrong 1`] = `
Object {
  "Authorization::4923fabc-aab2-4804-b92b-2aa96c4999a1": undefined,
}
`;

exports[`validateAuthorization should throw error when deviceId is wrong 1`] = `
Object {
  "Authorization::4923fabc-aab2-4804-b92b-2aa96c4999a1": undefined,
}
`;

exports[`validateAuthorization should throw error when email is wrong 1`] = `
Object {
  "Authorization::4923fabc-aab2-4804-b92b-2aa96c4999a1": undefined,
}
`;

exports[`validateAuthorization should throw error when grantType is wrong 1`] = `
Object {
  "Authorization::4923fabc-aab2-4804-b92b-2aa96c4999a1": undefined,
}
`;
